tools/tplg_parser/tokens.c

```diff
@@ -49,7 +49,7 @@ int get_token_uint32_t(void *elem, void *object, uint32_t offset,
 		       uint32_t size)
 {
 	struct snd_soc_tplg_vendor_value_elem *velem = elem;
-	uint32_t *val = object + offset;
+	uint32_t *val = (uint32_t *)((uint8_t *)object + offset);
```

Comment:
In the larger #6513 , @lyakh noted that `(uint32_t *)((uint8_t *)object + offset)` can cause unaligned access.

How about `(uint32_t *)object + offset/4` instead to cause a crash if offset is not aligned?

Or maybe an `assert()`

Reply:
@marc-hb yes, but I didn't imply which way was preferred. Maybe the unaligned access was a feature and not a bug. Although I *think* Xtensa doesn't like them, but I might be wrong.

Reply:
The code in kernel is like the patch above. See sound/soc/sof/topology.c line 346. The usage in tplg_parser is similar.

Some assume aligned macro could be used?

Reply:
Lets keep it simple and move on. Non alignment needs to be supported due to packing.
```diff
diff --git a/tools/tplg_parser/tokens.c b/tools/tplg_parser/tokens.c
index b85256d8f..c728b50e9 100644
--- a/tools/tplg_parser/tokens.c
+++ b/tools/tplg_parser/tokens.c
@@ -49,7 +49,8 @@ int get_token_uint32_t(void *elem, void *object, uint32_t offset,
                       uint32_t size)
 {
        struct snd_soc_tplg_vendor_value_elem *velem = elem;
-       uint32_t *val = object + offset;
+       char *vobject = object;
+       uint32_t *val = (uint32_t *)(vobject + offset);
 
        *val = velem->value;
        return 0;
```

Reply:
@lgirdwood but does Xtensa even support unaligned access? I see this in the list of exception causes:
```
#define EXCCAUSE_UNALIGNED		9	/* Unaligned Load or Store */
```
and I think when we pack data for IPC or similar ABIs we make sure ourselves that all fields have suitable alignment

Reply:
This is topology and not parsed by running FW, simulator can trap or enable this option (but we've never hit so I suspect CC is not directly loading the cast, but doing a shift load sequence).

Reply:
I followed Liam's proposal and updated all `uint32_t *val = (uint32_t *)((uint8_t *)object + offset);` occurrences. Hope it's OK now.

Reply:
> This is topology and not parsed by running FW, simulator can trap or enable this option (but we've never hit so I suspect CC is not directly loading the cast, but doing a shift load sequence).

@lgirdwood aah, right, this is never sent to the firmware as is, sorry for confusion